Update custom op API

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #106088
* #106076
* #106075
* __->__ #105947
* #106036
* #106035

As described in
https://docs.google.com/document/d/1aGWtgxV3HppuxQAdddyPrs74_aEntpkYt9MalnCKnhk/edit

This PR changes the CustomOp API to be private and adds new public
wrappers around it so that the user does not need to know about the
"CustomOp" object. We've effectively changed the "CustomOp" object to be
some metadata about the operator that the user does not directly
interact with.

The "updated custom op API" is in torch._custom_ops. Pending good customer
feedback, we will promote this module to torch.custom_ops.

NB: I cannot move around the older torch._custom_op APIs yet because
people are already using them.

Test Plan:
- I changed all of our tests to use the new `torch._custom_ops` module
instead of the old CustomOp API.

zou3519 committed:
Update custom op API

As described in
https://docs.google.com/document/d/1aGWtgxV3HppuxQAdddyPrs74_aEntpkYt9MalnCKnhk/edit

This PR changes the CustomOp API to be private and adds new public
wrappers around it so that the user does not need to know about the
"CustomOp" object. We've effectively changed the "CustomOp" object to be
some metadata about the operator that the user does not directly
interact with.

The "udated custom op API" is in torch._custom_ops. Pending good customer
feedback, we will promote this module to torch.custom_ops.

NB: I cannot move around the older torch._custom_op APIs yet because
people are already using them.

Test Plan:
- I changed all of our tests to use the new `torch._custom_ops` module
instead of the old CustomOp API.

[ghstack-poisoned]

zou3519 committed:
Update on "Update custom op API"


As described in
https://docs.google.com/document/d/1aGWtgxV3HppuxQAdddyPrs74_aEntpkYt9MalnCKnhk/edit

This PR changes the CustomOp API to be private and adds new public
wrappers around it so that the user does not need to know about the
"CustomOp" object. We've effectively changed the "CustomOp" object to be
some metadata about the operator that the user does not directly
interact with.

The "udated custom op API" is in torch._custom_ops. Pending good customer
feedback, we will promote this module to torch.custom_ops.

NB: I cannot move around the older torch._custom_op APIs yet because
people are already using them.

Test Plan:
- I changed all of our tests to use the new `torch._custom_ops` module
instead of the old CustomOp API.

[ghstack-poisoned]

zou3519 committed:
Update on "Update custom op API"


As described in
https://docs.google.com/document/d/1aGWtgxV3HppuxQAdddyPrs74_aEntpkYt9MalnCKnhk/edit

This PR changes the CustomOp API to be private and adds new public
wrappers around it so that the user does not need to know about the
"CustomOp" object. We've effectively changed the "CustomOp" object to be
some metadata about the operator that the user does not directly
interact with.

The "updated custom op API" is in torch._custom_ops. Pending good customer
feedback, we will promote this module to torch.custom_ops.

NB: I cannot move around the older torch._custom_op APIs yet because
people are already using them.

Test Plan:
- I changed all of our tests to use the new `torch._custom_ops` module
instead of the old CustomOp API.

[ghstack-poisoned]

zou3519 committed:
Update on "Update custom op API"


As described in
https://docs.google.com/document/d/1aGWtgxV3HppuxQAdddyPrs74_aEntpkYt9MalnCKnhk/edit

This PR changes the CustomOp API to be private and adds new public
wrappers around it so that the user does not need to know about the
"CustomOp" object. We've effectively changed the "CustomOp" object to be
some metadata about the operator that the user does not directly
interact with.

The "updated custom op API" is in torch._custom_ops. Pending good customer
feedback, we will promote this module to torch.custom_ops.

NB: I cannot move around the older torch._custom_op APIs yet because
people are already using them.

Test Plan:
- I changed all of our tests to use the new `torch._custom_ops` module
instead of the old CustomOp API.

[ghstack-poisoned]

zou3519 committed:
Update on "Update custom op API"


As described in
https://docs.google.com/document/d/1aGWtgxV3HppuxQAdddyPrs74_aEntpkYt9MalnCKnhk/edit

This PR changes the CustomOp API to be private and adds new public
wrappers around it so that the user does not need to know about the
"CustomOp" object. We've effectively changed the "CustomOp" object to be
some metadata about the operator that the user does not directly
interact with.

The "updated custom op API" is in torch._custom_ops. Pending good customer
feedback, we will promote this module to torch.custom_ops.

NB: I cannot move around the older torch._custom_op APIs yet because
people are already using them.

Test Plan:
- I changed all of our tests to use the new `torch._custom_ops` module
instead of the old CustomOp API.

[ghstack-poisoned]

zou3519 committed:
Update on "Update custom op API"


As described in
https://docs.google.com/document/d/1aGWtgxV3HppuxQAdddyPrs74_aEntpkYt9MalnCKnhk/edit

This PR changes the CustomOp API to be private and adds new public
wrappers around it so that the user does not need to know about the
"CustomOp" object. We've effectively changed the "CustomOp" object to be
some metadata about the operator that the user does not directly
interact with.

The "updated custom op API" is in torch._custom_ops. Pending good customer
feedback, we will promote this module to torch.custom_ops.

NB: I cannot move around the older torch._custom_op APIs yet because
people are already using them.

Test Plan:
- I changed all of our tests to use the new `torch._custom_ops` module
instead of the old CustomOp API.

[ghstack-poisoned]

soulitzer commented on torch/_custom_ops.py:
```diff
@@ -0,0 +1,327 @@
+
+    Also see
+    https://docs.google.com/document/d/1aGWtgxV3HppuxQAdddyPrs74_aEntpkYt9MalnCKnhk
+    for more details.
```
```suggestion
    for a detailed guide on how to define a custom op
```

soulitzer replied:
(in some other places too)

soulitzer commented on torch/_custom_ops.py:
```diff
@@ -0,0 +1,327 @@
+
+    def inner(func):
+        custom_op = _find_custom_op(qualname)
+        custom_op.impl_backward(output_differentiability, _stacklevel=3)(func)
```
`return func`

soulitzer commented on torch/_custom_ops.py:
```diff
@@ -0,0 +1,327 @@
+    if func_or_schema is None:
+        return inner
+    if isinstance(func_or_schema, str):
+        return _custom_op_with_schema(qualname, func_or_schema)
```
It looks like we (officially?) support this use case as well? I don't see any examples demonstrating it though.
```
>>> @torch._custom_ops.custom_op("mylibrary::numpy_sin")
>>> def numpy_sin(x: Tensor) -> Tensor:
>>>     raise NotImplementedError()
>>> 
>>> # register some impls
>>> 
>>> out = numpy_sin(input) 
```
Would it make reading code involving custom operators clearer to read ifwe force a single way of doing things, i.e. require users to call into `torch.ops.aten` themselves?

zou3519 replied:
We should require users to call into torch.ops.aten themselves. I wasn't really sure what to return with this decorator. I could return the original function?

soulitzer replied:
Aha, if we return the original function, then the error there would actually come in handy!

zou3519 replied:
Returned the original function

zou3519 commented on test/test_custom_ops.py:
```diff
@@ -8,10 +8,12 @@
 from functorch import make_fx
 from torch import Tensor
-from torch._custom_op.impl import custom_op
+from torch._custom_op.impl import custom_op, CustomOp
```
We don't need to deprecate it because it is not public API. I will make a cleanup pass sometime in the future because there are existing users who use the APIs. Pointing to the new API is a good suggestion, I will do that in this PR.

zou3519 commented on test/test_custom_ops.py:
```diff
@@ -361,39 +363,34 @@ def test_invalid_schemas(self):
+            @custom_ops.custom_op(f"{TestCustomOp.test_ns}::foo")
             def baz(x: Tensor) -> Tensor:
-                ...
+                raise NotImplementedError()
```
The alternative for this is that we ask the user to write code that looks like the following
```
def baz(x: Tensor) -> Tensor:
    ...  # noqa: F111
```
(or whatever the number is). There is no way for us (pytorch, the framework) to disable the mypy lint rule in user code.

Suggesting the user to ignore a mypy lint rule doesn't sound like a good user experience.

zou3519 replied:
Ehh I agree that NotImplementedError() looks weird though, because you'd expect someone to fill in the blanks. Maybe it should be a different type of error?

soulitzer replied:
Hmmm maybe? I don't have any better ideas though at the moment.

zou3519 committed:
Update on "Update custom op API"


As described in
https://docs.google.com/document/d/1aGWtgxV3HppuxQAdddyPrs74_aEntpkYt9MalnCKnhk/edit

This PR changes the CustomOp API to be private and adds new public
wrappers around it so that the user does not need to know about the
"CustomOp" object. We've effectively changed the "CustomOp" object to be
some metadata about the operator that the user does not directly
interact with.

The "updated custom op API" is in torch._custom_ops. Pending good customer
feedback, we will promote this module to torch.custom_ops.

NB: I cannot move around the older torch._custom_op APIs yet because
people are already using them.

Test Plan:
- I changed all of our tests to use the new `torch._custom_ops` module
instead of the old CustomOp API.

[ghstack-poisoned]

zou3519 committed:
Update on "Update custom op API"


As described in
https://docs.google.com/document/d/1aGWtgxV3HppuxQAdddyPrs74_aEntpkYt9MalnCKnhk/edit

This PR changes the CustomOp API to be private and adds new public
wrappers around it so that the user does not need to know about the
"CustomOp" object. We've effectively changed the "CustomOp" object to be
some metadata about the operator that the user does not directly
interact with.

The "updated custom op API" is in torch._custom_ops. Pending good customer
feedback, we will promote this module to torch.custom_ops.

NB: I cannot move around the older torch._custom_op APIs yet because
people are already using them.

Test Plan:
- I changed all of our tests to use the new `torch._custom_ops` module
instead of the old CustomOp API.

[ghstack-poisoned]

zou3519 committed:
Update on "Update custom op API"


As described in
https://docs.google.com/document/d/1aGWtgxV3HppuxQAdddyPrs74_aEntpkYt9MalnCKnhk/edit

This PR changes the CustomOp API to be private and adds new public
wrappers around it so that the user does not need to know about the
"CustomOp" object. We've effectively changed the "CustomOp" object to be
some metadata about the operator that the user does not directly
interact with.

The "updated custom op API" is in torch._custom_ops. Pending good customer
feedback, we will promote this module to torch.custom_ops.

NB: I cannot move around the older torch._custom_op APIs yet because
people are already using them.

Test Plan:
- I changed all of our tests to use the new `torch._custom_ops` module
instead of the old CustomOp API.

[ghstack-poisoned]

soulitzer reviewed (approved):
LGTM

zou3519 commented:
@pytorchbot merge

vadimkantorov commented:
@zou3519 related ideas on custom ops examples: https://github.com/pytorch/pytorch/issues/68066#issuecomment-1611836317

zou3519 commented:
> also interesting would be examples on how to register ops currently missing in pytorch, e.g. adding a quantized impl or some missing sparse op, so that dispatcher would use it and let go the exception that some op impl can't be found

This isn't a currently supported mode yet

> also interesting an example of adding a custom backward impl via a decorator

This is coming soon!

> also interesting an example of cuda op, but without having to manually add a C++ launcher stub (e.g. so that pytorch would generate the kernel launcher and maybe that the kernel can be compiled by nvrtc - like in CuPy)

This is interesting, but I'm not sure it is much of a user experience improvement. Are you saying you want PyTorch to auto-generate and compile the C++ launcher stub?

vadimkantorov commented:
> This isn't a currently supported mode yet
E.g. would be cool to add my third-party example impl of `quantized.linear_dynamic` for CUDA tensor with just a decorator / or registry call like

```python
@torch.custom_ops.register_impl("torch::quantized::linear_dynamic", "CudaTensor", "forward")
def my_impl_quantized_linear_dynamic(...):
    ...
```

This would make experimentation simpler for funny backends like NestedTensors or structured linear operators and so forth. And same for more specialized optimized impls like int8 (to avoid upcasts)


> This is interesting, but I'm not sure it is much of a user experience improvement. Are you saying you want PyTorch to auto-generate and compile the C++ launcher stub?

Yeah, like in cupy. So that we only need to provide a `*.cu` cuda code - it saves boilerplate if we directly pass in pointers and ints. And compiling with nvrtc is faster than nvcc I think.
